### PR TITLE
bluetooth: host: smp: Add bondable flag overlay per connection

### LINKED ISSUE #57070: Bluetooth: Host: SMP: set bondable mode per identity or connection object
The current API for changing the bondable mode uses the global flag:

```void bt_set_bondable(bool enable);```

With Zephyr support for multiple Bluetooth identities, the API for changing the bondable mode should be more fine-grained and allow for setting the bondable mode for a specific Bluetooth identity or connection object that is linked with the identity.

```void bt_set_bondable(uint8_t id, bool enable);```
or
```void bt_set_bondable(struct bt_conn *conn, bool enable);```
(used in the `connected` callback context)

The bondable requirements of one identity should not have an impact on another identity which can have a different set of requirements.

### PR DESCRIPTION
**Pull request by mkapala-nordic**
The current API for changing the bondable mode uses the global flag. With Zephyr support for multiple Bluetooth identities, the API for changing the bondable mode should be more fine-grained. The bondable requirements of one identity should not have an impact on another identity which can have a different set of requirements. This change introduces function to overlay bondable flag per connection.

Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/57070

### CODE REVIEW

**Comment by mkapala-nordic** [comment]
CC: @kapi-no @MarekPieta

File: subsys/bluetooth/host/smp.c [hunk 1]
```diff
@@ -288,6 +291,11 @@ static K_SEM_DEFINE(sc_local_pkey_ready, 0, 1);
  */
 #define BT_SMP_AUTH_CB_UNINITIALIZED	((atomic_ptr_val_t)bt_smp_pool)
 
+/* Value used to mark that bondable flag is not initialized.
```
Comment (MarekPieta) [inline subsys/bluetooth/host/smp.c:294]: per-connection bondable flag?

**Review by MarekPieta — APPROVED** [review]

**Review by kapi-no — APPROVED** [review]
Looking good (apart from small doc improvements suggested by @MarekPieta)

**Comment by kapi-no** [comment]
The naming convention and the implementation for this new `bt_bondable_overlay` API are similar to the `bt_smp_auth_cb_overlay` API function.

**Comment by mkapala-nordic** [comment]
Adressed @MarekPieta comments.

**Review by MarekPieta — APPROVED** [review]

**Review by kapi-no — APPROVED** [review]

File: include/zephyr/bluetooth/conn.h [hunk 2]
```diff
@@ -1081,6 +1081,25 @@ void bt_conn_cb_register(struct bt_conn_cb *cb);
...
+ *
+ *  If the bonding flag is not overlaid, the value will depend on global
+ *  configuration which is set using bt_set_bondable.
+ *  The default value of the global configuration is defined using
+ *  CONFIG_BT_BONDABLE Kconfig option.
+ *
+ *  @param conn Connection object.
+ *  @param enable Value allowing/disallowing to be bondable.
+ */
+int bt_bondable_overlay(struct bt_conn *conn, bool enable);
```
Comment (jhedberg) [inline include/zephyr/bluetooth/conn.h:1101]: I feel like `bt_conn_set_bondable()` would be the more appropriate name. I know we have `bt_smp_auth_cb_overlay()` as an internal API, but public APIs operating on a `struct bt_conn *` should preferably be prefixed with `bt_conn_`.

Comment (mkapala-nordic) [reply include/zephyr/bluetooth/conn.h]: Agreed, name changed.

**Review by kapi-no — APPROVED** [review]

**Review by jhedberg — DISMISSED** [review]

**Review by MarekPieta — APPROVED** [review]

**Review by jori-nordic — CHANGES_REQUESTED** [review]
could you add a bsim test checking the new functionality?
and also mark this new API as experimental (in case we later find some better way to achieve the same thing)

**Comment by mkapala-nordic** [comment]
Marked as experimental + added bsim test as per @jori-nordic request.

File: subsys/bluetooth/host/Kconfig [hunk 3]
```diff
@@ -600,6 +600,13 @@ config BT_BONDING_REQUIRED
 	  set the bondable flag in their pairing request. Any other kind of
 	  requests will be rejected.
 
+config BT_BONDABLE_PER_CONNECTION
+	bool "Set/clear the bonding flag per-connection [EXPERIMENTAL]"
+	select EXPERIMENTAL
+	help
+	  Enable support for setting/clearing the bonding flag on a
+	  per-connection basis.
```
Comment (kapi-no) [inline subsys/bluetooth/host/Kconfig:608]:
```suggestion
	  Enable support for the bt_conn_set_bondable API function that is 
	  used to set/clear the bonding flag on a per-connection basis.
```

**Review by kapi-no — APPROVED** [review]

**Review by MarekPieta — APPROVED** [review]
Change in host looks good

File: tests/bsim/bluetooth/host/security/bond_per_connection/src/bs_bt_utils.c [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+	}
+
+	if (err != 0) {
+		clear_g_conn();
+		return;
+	}
+
+	SET_FLAG(flag_is_connected);
+
+	if (bt_conn_set_bondable(conn, bondable)) {
```
Comment (jori-nordic) [inline tests/bsim/bluetooth/host/security/bond_per_connection/src/bs_bt_utils.c:65]: could you add another round of bonding without calling `bt_conn_set_bondable`?
like have a bondable by default device, try bonding and see that it works and then continue with the two test rounds you already have

Comment (mkapala-nordic) [reply tests/bsim/bluetooth/host/security/bond_per_connection/src/bs_bt_utils.c]: Added such case.

**Review by jhedberg — DISMISSED** [review]

**Review by jori-nordic — APPROVED** [review]

**Review by alwa-nordic — APPROVED** [review]

**Comment by carlescufi** [comment]
@jhedberg would you mind taking another look?

**Review by jhedberg — APPROVED** [review]

**Pull request merged**